# table of contents
1. [intro](#intro)
2. [dataset](#dataset)
3. [dictionary](#dictionary)
4. [exploration](#exploration)
5. [cleaning](#cleaning)
6. [statistical summary](#statistical-summary)

# intro
A chain of hotels has just hired me as a data analyst. They have noticed that the cancellation rate has risen in the past few years. This often leads to rooms that are left unrented for multiple days at a time.

Hotel management is interested in developing a model to predict the likelihood that a customer will cancel their reservation. If successful, this could be used to optimize their booking service and anticipate when cancellations will occur.

I'm going to prepare a report that is accessible to a broad audience. It should outline my motivation, steps, findings, and conclusions.

# dataset
This dataset consists of booking data from a city hotel and a resort hotel.\
It includes many details about the bookings, including room specifications, the length of stay, the time between the booking and the stay, whether the booking was canceled, and how the booking was made.

# dictionary
_Note: For binary variables: `1` = true and `0` = false._

| Column                                                                                                                                                                                                          | Explanation                                                                                                                            |   |   |   |
|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------|---|---|---|
| is_canceled                                                                                                                                                                                                     | Binary variable indicating whether a booking was canceled                                                                              |   |   |   |
| lead_time                                                                                                                                                                                                       | Number of days between booking date and arrival date                                                                                   |   |   |   |
| arrival_date_week_number, arrival_date_day_of_month, arrival_date_month                                                                                                                                         | Week number, day date, and month number of arrival date                                                                                |   |   |   |
| stays_in_weekend_nights, stays_in_week_nights                                                                                                                                                                   | Number of weekend nights (Saturday and Sunday) and weeknights (Monday to Friday) the customer booked                                   |   |   |   |
| adults, children, babies                                                                                                                                                                                        | Number of adults, children, babies booked for the stay                                                                                 |   |   |   |
| is_repeated_guest                                                                                                                                                                                               | Binary variable indicating whether the customer was a repeat guest                                                                     |   |   |   |
| previous_cancellations                                                                                                                                                                                          | Number of prior bookings that were canceled by the customer                                                                            |   |   |   |
| previous_bookings_not_canceled                                                                                                                                                                                  | Number of prior bookings that were not canceled by the customer                                                                        |   |   |   |
| required_car_parking_spaces                                                                                                                                                                                     | Number of parking spaces requested by the customer                                                                                     |   |   |   |
| total_of_special_requests                                                                                                                                                                                       | Number of special requests made by the customer                                                                                        |   |   |   |
| avg_daily_rate                                                                                                                                                                                                  | Average daily rate, as defined by dividing the sum of all lodging transactions by the total number of staying nights                   |   |   |   |
| booked_by_company                                                                                                                                                                                               | Binary variable indicating whether a company booked the booking                                                                        |   |   |   |
| booked_by_agent                                                                                                                                                                                                 | Binary variable indicating whether an agent booked the booking                                                                         |   |   |   |
| hotel_City                                                                                                                                                                                                      | Binary variable indicating whether the booked hotel is a "City Hotel"                                                                  |   |   |   |
| hotel_Resort                                                                                                                                                                                                    | Binary variable indicating whether the booked hotel is a "Resort Hotel"                                                                |   |   |   |
| meal_BB                                                                                                                                                                                                         | Binary variable indicating whether a bed & breakfast meal was booked                                                                   |   |   |   |
| meal_HB                                                                                                                                                                                                         | Binary variable indicating whether a half board meal was booked                                                                        |   |   |   |
| meal_FB                                                                                                                                                                                                         | Binary variable indicating whether a full board meal was booked                                                                        |   |   |   |
| meal_No_meal                                                                                                                                                                                                    | Binary variable indicating whether there was no meal package booked                                                                    |   |   |   |
| market_segment_Aviation, market_segment_Complementary, market_segment_Corporate, market_segment_Direct, market_segment_Groups, market_segment_Offline_TA_TO, market_segment_Online_TA, market_segment_Undefined | Indicates market segment designation with a value of 1. "TA"= travel agent, "TO"= tour operators                                       |   |   |   |
| distribution_channel_Corporate, distribution_channel_Direct, distribution_channel_GDS, distribution_channel_TA_TO, distribution_channel_Undefined                                                               | Indicates booking distribution channel with a value of 1. "TA"= travel agent, "TO"= tour operators, "GDS" = Global Distribution System |   |   |   |
| reserved_room_type_A, reserved_room_type_B, reserved_room_type_C, reserved_room_type_D, reserved_room_type_E, reserved_room_type_F, reserved_room_type_G, reserved_room_type_H, reserved_room_type_L            | Indicates code of room type reserved with a value of 1. Code is presented instead of designation for anonymity reasons                 |   |   |   |
| deposit_type_No_Deposit                                                                                                                                                                                         | Binary variable indicating whether a deposit was made                                                                                  |   |   |   |
| deposit_type_Non_Refund                                                                                                                                                                                         | Binary variable indicating whether a deposit was made in the value of the total stay cost                                              |   |   |   |
| deposit_type_Refundable                                                                                                                                                                                         | Binary variable indicating whether a deposit was made with a value under the total stay cost                                           |   |   |   |
| customer_type_Contract                                                                                                                                                                                          | Binary variable indicating whether the booking has an allotment or other type of contract associated to it                             |   |   |   |
| customer_type_Group                                                                                                                                                                                             | Binary variable indicating whether the booking is associated to a group                                                                |   |   |   |
| customer_type_Transient                                                                                                                                                                                         | Binary variable indicating whether the booking is not part of a group or contract, and is not associated to other transient booking    |   |   |   |
| customer_type_Transient-Party                                                                                                                                                                                   | Binary variable indicating whether the booking is transient, but is associated to at least another transient booking                   |   |   |   |

[Source](https://www.kaggle.com/jessemostipak/hotel-booking-demand/) and [license](https://creativecommons.org/licenses/by/4.0/) of data. 

**Citation**: The data is originally from an article called [Hotel booking demand datasets](https://www.sciencedirect.com/science/article/pii/S2352340918315191) by Nuno Antonio, Ana de Almeida, and Luis Nunes. It was cleaned by Thomas Mock and Antoine Bichat for [#TidyTuesday during the week of February 11th, 2020](https://github.com/rfordatascience/tidytuesday/blob/master/data/2020/2020-02-11/readme.md).

# exploration

In [1]:
import pandas as pd
df = pd.read_csv("../data/01_raw_data.csv")
print(df.shape)
display(df.head(61))
display(df.tail(61))

(119210, 53)


is_canceled  lead_time  arrival_date_week_number  \
0             0        342                        27   
1             0        737                        27   
2             0          7                        27   
3             0         13                        27   
4             0         14                        27   
..          ...        ...                       ...   
56            0          0                        27   
57            0          0                        27   
58            0          0                        27   
59            0         14                        27   
60            0         10                        27   

    arrival_date_day_of_month  arrival_date_month  stays_in_weekend_nights  \
0                           1                   7                        0   
1                           1                   7                        0   
2                           1                   7                        0   
3                           1                   7                        0   
4                           1                   7                        0   
..                        ...                 ...                      ...   
56                          2                   7                        0   
57                          2                   7                        0   
58                          2                   7                        0   
59                          2                   7                        0   
60                          2                   7                        0   

    stays_in_week_nights  adults  children  babies  ...  reserved_room_type_G  \
0                      0       2       0.0       0  ...                     0   
1                      0       2       0.0       0  ...                     0   
2                      1       1       0.0       0  ...                     0   
3                      1       1       0.0       0  ...                     0   
4                      2       2       0.0       0  ...                     0   
..                   ...     ...       ...     ...  ...                   ...   
56                     1       2       0.0       0  ...                     0   
57                     1       2       0.0       0  ...                     0   
58                     1       2       0.0       0  ...                     1   
59                     2       2       0.0       0  ...                     0   
60                     2       2       0.0       0  ...                     1   

    reserved_room_type_H  reserved_room_type_L  deposit_type_No_Deposit  \
0                      0                     0                        1   
1                      0                     0                        1   
2                      0                     0                        1   
3                      0                     0                        1   
4                      0                     0                        1   
..                   ...                   ...                      ...   
56                     1                     0                        1   
57                     0                     0                        1   
58                     0                     0                        1   
59                     0                     0                        1   
60                     0                     0                        1   

    deposit_type_Non_Refund  deposit_type_Refundable  customer_type_Contract  \
0                         0                        0                       0   
1                         0                        0                       0   
2                         0                        0                       0   
3                         0                        0                       0   
4                         0                        0                       0   
..                      ...             

is_canceled  lead_time  arrival_date_week_number  \
119149            0         30                        35   
119150            0        177                        35   
119151            0        104                        35   
119152            0         27                        35   
119153            0        231                        35   
...             ...        ...                       ...   
119205            0         23                        35   
119206            0        102                        35   
119207            0         34                        35   
119208            0        109                        35   
119209            0        205                        35   

        arrival_date_day_of_month  arrival_date_month  \
119149                         31                   8   
119150                         29                   8   
119151                         31                   8   
119152                         31                   8   
119153                         28                   8   
...                           ...                 ...   
119205                         30                   8   
119206                         31                   8   
119207                         31                   8   
119208                         31                   8   
119209                         29                   8   

        stays_in_weekend_nights  stays_in_week_nights  adults  children  \
119149                        0                     3       2       0.0   
119150                        0                     5       2       0.0   
119151                        0                     3       1       0.0   
119152                        0                     3       2       0.0   
119153                        1                     5       2       0.0   
...                         ...                   ...     ...       ...   
119205                        2                     5       2       0.0   
119206                        2                     5       3       0.0   
119207                        2                     5       2       0.0   
119208                        2                     5       2       0.0   
119209                        2                     7       2       0.0   

        babies  ...  reserved_room_type_G  reserved_room_type_H  \
119149       0  ...                     0                     0   
119150       0  ...                     0                     0   
119151       0  ...                     0                     0   
119152       0  ...                     0                     0   
119153       0  ...                     0                     0   
...        ...  ...                   ...                   ...   
119205       0  ...                     0                     0   
119206       0  ...                     0                     0   
119207       0  ...                     0                     0   
119208       0  ...                     0                     0   
119209       0  ...                     0                     0   

        reserved_room_type_L  deposit_type_No_Deposit  \
119149                     0                        1   
119150                     0                        1   
119151                     0                        1   
119152                     0                        1   
119153                     0                        1   
...                      ...                      ...   
119205                     0                        1   
119206                     0                        1   
119207                     0                        1   
119208                     0                        1   
119209                     0                        1   

        deposit_type_Non_Refund  deposit_type_Refundable  \
119149                        0                        0   
119150                        0                        0   
119151                        0

In [ ]:
"""
By default, panda methods head and tail give preview of maximum 20 columns, 10 from each side of a dataframe.
The current df has 53 columns, so the default preview doesn't suffice.
The following setting should adjust this and return the preview of all the 53 columns.
"""
pd.set_option('display.max_columns', None)
display(df.head(61))
display(df.tail(61))

In [ ]:
# "children" values look like floats. is dtype float? why?
df.info()

# cleaning

In [ ]:
#df["children"] = df["children"].astype(int)
#df
# Attempt to immediately convert "children" column dtype to integer (before I forget) fails, due to NaN values.
# So, let's deal with the NaNs.

In [ ]:
df.isna().sum()

# "children" is the only column with NaN values, totally 4.

In [ ]:
# 4 NaN values / 119210 entries is a negligent proportion, so whatever we do with NaN is irrelevant. Let's drop them.

df = df.dropna()
df

In [ ]:
for column in df.columns:
    print(df[column].value_counts())
    print()

In [ ]:
# Drop the columns which give no information.
df = df.drop(["market_segment_Undefined", "distribution_channel_Undefined"], axis=1)
df

In [ ]:
df["children"] = df["children"].astype(int)
df.info()

# Now all the data is in an appropriate type. Let's check for duplicates.

In [ ]:
df.duplicated().any()
#There are duplicates. How many?

In [ ]:
df.duplicated().sum()
# There are 36024 duplicates. Where are they?

In [ ]:
df[df.duplicated(keep=False)]

1/3 of rows (45039/119210) in the df are duplicated.\
Is this a problem, though? Maybe, but it's hard to know.\
This df is expected to have duplicated rows.\
More than half of the columns are categorical variables with binary values (1,0).\
Out of 119210 entries/guests, it's normal that many guests had identical preferences and booking choices.\
So, what to do with these duplicates? Nothing.

# statistical summary

In [ ]:
df.describe()

Of **119206 bookings**:
- **37% were canceled**
- rooms were **booked 2 (median) - 3,5 (mean) months ahead**
  - median < mean = more people tend to book closer to the stay-date
  - but some far ahead bookings (2 years) drag the distribution tail to the right
- **peak arrivals** are **mid-june to mid-july**, showing by the:
  - arrival_date_week_number: 27 (mean) - 28 (median),
  - arrival_date_day_of_month: 15 
  - arrival_date_month: 6.6 (mean) - 7 (median)
- bookings include at least **1 weekend day and/or 2 weekdays**
- bookings were made for **2 adults**, no children or babies
- **3% repeated guests**
- **no history of cancelations**, which aligns with 97% being first-timers
- **car parking space not required**
- **no special requests**
- **daily rate: 95 (median) - 102 (mean)**
- **86% agent** bookings, 6% company bookings
- **66% city hotel** bookings, 34% resort bookings
- meals: **77% bed&breakfast**, 12% half-breakfast, 10% no meal, 1% full breakfast
- market segment: **47% online travel agent (TA)**, 20% offline TA, 17% groups, 11% direct, 4% corporate, 1% other segments
- distribution channel: **82% TA**, 12% direct, 5% corporate
- **72% reserved rooms were type A**, 16% D, 5% E, 7% other 6 types
- **88% booked without deposit**, 12% with a deposit
- **75% individual (transient) bookings**, 21% transient and associated to another transient booking, 4% other arrangements.

**Outliers:**
- 2 years lead time
- 19 weekends
- 50 days
- 55 adults
- 10 children
- 10 babies
- 26 cancelations
- 8 car parking spaces
- 5 special requests
-daily rate: -6.38
- daily rate: 5400

In [3]:
df.to_csv("../data/02_cleaned_data.csv", index=False)

Next: notebook_02_univariate_analysis